README.md
# alphatunes

AlphaTunes is a free music app partly developed on livestream at my youtube channel 'youtube.com/alphascript'.

## Installation

Head to my website: 'coming soon...' and follow the installation instructions there.

## Usage

Head to the website and download the standalone .exe. ~ coming soon

## Contribution

If you have any feature requests, feel free to let me know.
If you want to fix any bugs, feel free to make a pull request. 
If you have any bug reports, please let me know.

## License

There is no license selected, this is because **I DO NOT** want people to share this code as their own or release their own versions. The code is public as a learning resource for interested people, nothing more.
Please do not pirate/claim this project as your own, or even slightly modify it and claim it.

### Thanks for checking this out!


changelog.txt
23/11/2020 - 18:53 - FIXED - Canvas scrolling above window.
23/11/2020 - 17:30 - ADDED - Playlist class for loading, storing, creating, editing.
23/11/2020 - 20:00 - ADDED - Playlist display refresh function.
23/11/2020 - 20:36 - FIXED - Playlist loading function.
23/11/2020 - 20:45 - FIXED - Playlist update resize function to update scrollbar.
11/12/2020 - 12:53 - CHANGED - Storing songs now involves the url of the page and the thumbnail url.
11/12/2020 - 13:13 - ADDED - Method to check and create playlist directory if not found (automatically creates userdata if not found).

gui.py
import tkinter as tk
import tkinter.ttk as ttk
import sys
import os

# pages
from mainpage import MainPage

class GUI(tk.Tk):
    def __init__(self):
        tk.Tk.__init__(self)

        # colours
        self.bg_colour = "#202020"
        self.playlist_bg_colour = "#151515"
        self.taskbar_bg_colour = "#333333"
        self.sb_bar_colour = "#505050"
        self.sb_selected_colour = "#656565"
        self.dark_white_colour = "#8a8a8a"

        # fonts
        self.header_font = "Arial 32 bold"
        self.normal_font = "Arial 15" # font for playlist names and songs
        self.song_font = "Arial 12 bold"

        # images

        # basic setup
        self.default_width = 1000 # calculate this?
        self.default_height = 750 # calculate this?
        
        self.title("alphatunes")
        self.geometry(f"{self.default_width}x{self.default_height}")
        self.resizable(0, 0)
        self.minsize(self.default_width, self.default_height)

        # view mode (current frame)

        # frames, make resizable
        container = tk.Frame(self)
        container.place(x=0, y=0, width=self.default_width, height=self.default_height)
        self.frames = {"MAIN" : MainPage(container, self)}

        # call main frame
        self.frames["MAIN"].tkraise()

    def initialise_frame(self, frame_object, image=None): # for all the basic setup stuff in frames
        frame_object.place(x=0, y=0, width=self.default_width, height=self.default_height)
        frame_object.configure(bg=self.bg_colour)

        if image != None: # for making it easier to initialise a frame with a background
            background_gif = tk.PhotoImage(file=self.resource_path(image))
            background = tk.Label(frame_object, image=background_gif, borderwidth=0, highlightthickness=0)
            background.image = background_gif # keep reference to image
            background.place(x=0, y=0)

    def resource_path(self, relative_path): # copied from stackoverflow, for making the program a .exe in future
        """ Get absolute path to resource, works for dev and for PyInstaller """
        try:
            base_path = sys._MEIPASS # PyInstaller creates a temp folder and stores path in _MEIPASS, ignore error
            
        except Exception:
            base_path = os.path.abspath(".")

        return os.path.join(base_path, relative_path)        
        

        

        

    


player.py
import youtube_dl
import threading
import os

# player functions
def search(title, amount=10): # search for results, works
    info = youtube_dl.YoutubeDL({"quiet" : True}).extract_info(f"ytsearch{amount}:{title}", download=False, ie_key="YoutubeSearch")

    data = {}

    for entry in info["entries"]:
        data[entry["title"]] = (entry["webpage_url"], entry["thumbnails"][0]["url"]) # store title and thumbnail for display, url for downloads, keep this format for storing the songs in playlist too

    return data 

# wrapper for making background tasks
class BackgroundTask(threading.Thread): 
    def __init__(self, task, args=()): # arguments must be passed like a normal thread eg (arg1, arg2,)
        threading.Thread.__init__(self)
        self.task = task # the task to be executed
        self.args = args # task arguments
        self.alive = True # flag to control waiting for feedback
        self.feedback = None # to allow feedback if function returns

    def run(self):
        self.feedback = self.task(*self.args) # initiate the function and return data
        self.alive = False # kill thread and waiting loop

# playlist class, each playlist is its own instance
class Playlist:
    def __init__(self): # create function done in initialiser, basically use preset values
        self.directory = "/userdata/playlists"

        # playlist data
        self.name = f"Playlist {}"
        self.description = "..."
        self.songs = {}

    def validate_dirs(self):
        if not os.path.isdir(self.directory): # create folders if not found, check each time playlist is stored
            os.makedirs(self.directory)
            return True # validated

        return None # nothing done

    def store(self): # create or update a playlist
        validate_dirs()
        with open(os.path.join(self.directory, self.name, ".txt"), "w") as playlist:
            playlist.write(self.name + "\n")
            playlist.write(self.description + "\n")
            
            for song in self.songs:
                title = song
                url, thumbnail_url = self.songs[title]
                playlist.write(title + "\\" + url + "\\" + thumbnail_url + "\n")

            return True

    def load(self, playlist_name): # load a stored playlist
        if validate_dirs():
            return None # had to create file, there are no stored playlists
        
        files = os.listdir(self.directory)
        filename = playlist_name + ".txt"
        if filename not in files: # couldn't find file
            return None

        with open(os.path.join(self.directory, filename), "r") as playlist:
            data = playlist.readlines()
            if len(data) < 2: # the file must have a description and title, default at least, otherwise it is of an invalid format
                return False

            self.name = playlist_name
            self.description = data[0].strip("\n")

            for song in data[1:]:
                try:
                    parts = song.strip("\n").split("\\")
                    title = parts[0]
                    data = (parts[1], parts[2]) # make this metadata, include author, publish date?
                    self.songs[title] = data

                except IndexError: # invalid data format
                    return False

        return True
    
    def delete(self): # delete a playlist
        os.remove(self.directory, self.name + ".txt")
        del self


mainpage.py
import time
import tkinter as tk
import tkinter.ttk as ttk
import sys
import os
import player
import threading
import queue

class MainPage(tk.Frame):
    def __init__(self, parent, controller):
        #intialise
        tk.Frame.__init__(self, parent)
        self.controller = controller
        self.controller.initialise_frame(self)

        # settings
        self.current_hovered_widget = "" # store the previous colour of the element being hovered over
        self.search_results = [] # store the search results (to be displayed)
        self.search_result_widgets = [] # so that the search results can be updated

        # playlist information
        self.displayed_playlists = []
        self.loaded_playlist = None

        #styling
        self.style = ttk.Style()
        self.style.theme_use('clam')
        
        # custom scrollbars
        self.style.layout('custom.Vertical.TScrollbar', # create layout and remove the top/bottom arrows
             [('Vertical.Scrollbar.trough',
               {'children': [('Vertical.Scrollbar.thumb', 
                              {'expand': '1', 'sticky': 'nswe'})],
                'sticky': 'ns'})])
        
        self.style.layout('custom2.Vertical.TScrollbar', # create layout and remove the top/bottom arrows
             [('Vertical.Scrollbar.trough',
               {'children': [('Vertical.Scrollbar.thumb', 
                              {'expand': '1', 'sticky': 'nswe'})],
                'sticky': 'ns'})])
        
        self.style.configure("custom.Vertical.TScrollbar", gripcount=0, arrowsize=18, # custom colours (default arrowsize=15)
                background=self.controller.sb_bar_colour, darkcolor=self.controller.playlist_bg_colour, lightcolor=self.controller.playlist_bg_colour,
                troughcolor=self.controller.playlist_bg_colour, bordercolor=self.controller.playlist_bg_colour, arrowcolor=self.controller.playlist_bg_colour)

        self.style.configure("custom2.Vertical.TScrollbar", gripcount=0, arrowsize=18, # custom colours (default arrowsize=15)
                background=self.controller.sb_bar_colour, darkcolor=self.controller.bg_colour, lightcolor=self.controller.bg_colour,
                troughcolor=self.controller.bg_colour, bordercolor=self.controller.bg_colour, arrowcolor=self.controller.bg_colour)

        self.style.map("Vertical.TScrollbar", background=[("active", self.controller.sb_selected_colour)])
    
        # playlist section
        # playlist canvas, for the scrollbar areas (the canvas is its own frame)
        self.playlist_canvas = tk.Canvas(self, width=self.controller.default_width//5, height=(self.controller.default_height//12)*11, bg=self.controller.playlist_bg_colour, highlightthickness=0, borderwidth=0)
        self.playlist_canvas.place(x=17, y=0) # pushed 17px to the right for scrollbar, 18 has a weird white dot
        
        # playlist area
        self.playlist_area = tk.Frame(self.playlist_canvas)
        self.playlist_area.bind('<Configure>', self.resize_playlist_canvas_scroll) # resize the canvas scrollregion each time the size of the frame changes
        self.playlist_canvas.create_window(0, 0, window=self.playlist_area) # display frame inside the canvas
        
        # scrollbars
        self.playlist_scrollbar = ttk.Scrollbar(self, command=self.playlist_canvas_yview, orient="vertical", style="custom.Vertical.TScrollbar")
        self.playlist_scrollbar.place(x=0, y=0, height=(self.controller.default_height//12)*11) # set the scrollbar to the left of the screen
        self.playlist_canvas.configure(yscrollcommand=self.playlist_scrollbar.set) 
        
        # widgets
        playlist_title = tk.Label(self.playlist_canvas, text="PLAYLISTS", font=self.controller.normal_font, fg="white", bg=self.controller.playlist_bg_colour)
        self.playlist_canvas.create_window(self.controller.default_width//10-9, 10, window=playlist_title) # -9 for the scrollbar
        
        for i in range(50):
            label = tk.Label(self.playlist_canvas, text=f"Playlist {i}", font=self.controller.normal_font, cursor="hand2", fg=self.controller.dark_white_colour, bg=self.controller.playlist_bg_colour) # binding the widget to the canvas improves performance
            label.bind("<Enter>", lambda event, widget=label: self.highlight(event, widget))
            label.bind("<Leave>", lambda event, widget=label: self.unhighlight(event, widget))
            label.bind("<Button-1>", lambda event, playlist_title=label["text"]: self.display_playlist(event, playlist_title))
            self.playlist_canvas.create_window(self.controller.default_width//10-9, i*25+35, window=label) # -9 for the scrollbar
        
        create = tk.Label(self, text="Create Playlist", cursor="hand2", font=self.controller.normal_font, bg=self.controller.playlist_bg_colour, fg=self.controller.dark_white_colour, relief="sunken")
        create.bind("<Enter>", lambda event, widget=create: self.highlight(event, widget))
        create.bind("<Leave>", lambda event, widget=create: self.unhighlight(event, widget))
        
        create.place(x=0, y=((self.controller.default_height//12)*11), width=(self.controller.default_width//5)+17, height=self.controller.default_height-((self.controller.default_height//12)*11)) # have to re-calculate because of rounding which causes gaps
    
        # set scrollbar to default position at the top
        self.playlist_canvas.configure(scrollregion=self.playlist_canvas.bbox('all'))
        self.playlist_canvas.update_idletasks()
        self.playlist_canvas.yview_moveto(0)
        


        # main canvas
        self.main_canvas = tk.Canvas(self, width=self.controller.default_width-(self.controller.default_width//5)-34, height=(self.controller.default_height//12)*11, bg=self.controller.bg_colour, highlightthickness=0, borderwidth=0) # take away 34 for the scrollbars
        self.main_canvas.place(x=17+self.controller.default_width//5, y=-1) # pushed 17 to the right for scrollbar, -1 gets rid of white dot

        # main area
        self.main_area = tk.Frame(self.main_canvas)
        self.main_area.bind('<Configure>', self.resize_main_canvas_scroll) # resize the canvas scrollregion each time the size of the frame changes
        
        self.main_canvas.create_window(17+self.controller.default_width//5, 0, window=self.main_area) # display frame inside the canvas
        
        # scrollbars
        self.main_scrollbar = ttk.Scrollbar(self, command=self.main_canvas_yview, orient="vertical", style="custom2.Vertical.TScrollbar")
        self.main_scrollbar.place(x=self.controller.default_width-17, y=0, height=(self.controller.default_height//12)*11) # set the scrollbar to the right of the screen
        self.main_canvas.configure(yscrollcommand=self.main_scrollbar.set)

        # search mode
        search_label = tk.Label(self.main_canvas, text="Song Search:", fg="white", bg=self.controller.bg_colour, font=self.controller.header_font)
        self.main_canvas.create_window(self.calc_centre(self.main_canvas), 40, window=search_label)

        # search entry
        self.search_var = tk.StringVar()
        self.search_var.set("Search")
        self.search_entry = tk.Entry(self.main_canvas, textvariable=self.search_var, cursor="xterm", font=self.controller.normal_font)
        self.search_entry.bind("<Button-1>", lambda event: self.search_var.set(""))
        self.main_canvas.create_window(self.calc_centre(self.main_canvas), 100, window=self.search_entry)

        # search button
        self.search_button = tk.Button(self.main_canvas, text="Search", font=self.controller.normal_font, cursor="hand2", fg="green", bg=self.controller.playlist_bg_colour, command=lambda: self.song_search(self.search_var.get()))
        self.main_canvas.create_window(self.calc_centre(self.main_canvas), 150, window=self.search_button)
        
        # set scrollbar to default position at the top
        self.main_canvas.configure(scrollregion=self.main_canvas.bbox('all'))
        self.main_canvas.update_idletasks()
        self.main_canvas.yview_moveto(0)

        # task bar
        # canvas (to make display easier)
        self.taskbar_canvas = tk.Canvas(self, width=(self.controller.default_width//5)*4, height=self.controller.default_height//11, bg=self.controller.taskbar_bg_colour, highlightthickness=0, borderwidth=0)
        self.taskbar_canvas.place(x=17+self.controller.default_width//5, y=(self.controller.default_height//12)*11) # pushed 17 to the right for scrollbar, -1 gets rid of white dot

    def display_playlist(self, event, playlist_name): # display the playlist
        # load data
        self.loaded_playlist = player.Playlist()
        feedback = self.loaded_playlist.load(playlist_name)

        if feedback is None: # playlist could not be found
            print("playlist could not be found") # USE A MESSAGE BOX TO DISPLAY ERROR THEN UPDATE SCROLLBAR
            self.update_playlist_display()
            return

        elif not feedback: # invalid format
            print("invalid playlist format")
            pass

        self.main_canvas.delete("all") # clear main canvas
        
        # title
        playlist_title = tk.Label(self.main_canvas, text=playlist_name, fg="white", bg=self.controller.bg_colour, font=self.controller.header_font) # used for receiving the name of the playlist for writing
        self.main_canvas.create_window(self.calc_centre(self.main_canvas), 40, window=playlist_title)

        # display description
        playlist_description = tk.Label(self.main_canvas, text=self.loaded_playlist.description, bg=self.controller.bg_colour, fg=self.controller.dark_white_colour, font=self.controller.normal_font)
        self.main_canvas.create_window(self.calc_centre(self.main_canvas), 100, window=playlist_description)

        # display songs
        i = 0
        for song in self.loaded_playlist.songs:
            url, thumbnail_url = self.loaded_playlist.songs[song]
            song_label = tk.Label(self.main_canvas, text=song, bg=self.controller.bg_colour, fg=self.controller.dark_white_colour, font=self.controller.normal_font)
            self.main_canvas.create_window(self.calc_centre(self.main_canvas), 150+i*40, window=song_label)

    def update_playlist_display(self): # if a playlist is corrupt/deleted its going to have to refresh the display
        self.playlist_canvas.delete("all") # clear playlist canvas
        playlist_title = tk.Label(self.playlist_canvas, text="PLAYLISTS", font=self.controller.normal_font, fg="white", bg=self.controller.playlist_bg_colour)
        self.playlist_canvas.create_window(self.controller.default_width//10-9, 10, window=playlist_title) # -9 for the scrollbar
        
        playlists = os.listdir("userdata/playlists/") # find all the playlist files

        for i in range(len(playlists)):
            playlist = playlists[i]
            if not playlist.endswith(".txt"):
                continue

            name = playlist[:-4] # remove .txt
            self.displayed_playlists.append(name)

            display = tk.Label(self.playlist_canvas, text=name, font=self.controller.normal_font, cursor="hand2", fg=self.controller.dark_white_colour, bg=self.controller.playlist_bg_colour) # binding the widget to the canvas improves performance
            display.bind("<Enter>", lambda event, widget=display: self.highlight(event, widget))
            display.bind("<Leave>", lambda event, widget=display: self.unhighlight(event, widget))
            display.bind("<Button-1>", lambda event, playlist_title=name: self.display_playlist(event, playlist_title))
            self.playlist_canvas.create_window(self.controller.default_width//10-9, i*25+35, window=display) # -9 for the scrollbar

        self.resize_playlist_canvas_scroll(None) # resize the scrollbar for the updated playlists
        
    def create_playlist(self):
        # basically created a new playlist and store it, make sure to add to the list and use the normal functions
        # click on the title to change
        # click on the description to change
        # have a plus icon somewhere to add songs
        pass

    def song_search(self, title):
        if title.strip(" ") != "" and title.strip(" ") != "Search":
            task = player.BackgroundTask(player.search, args=(title,))
            task.start()

            while task.alive: # wait for feedback
                self.controller.update()
                time.sleep(0.01) # let CPU rest

            self.search_results = task.feedback

            # display results
            for result in self.search_result_widgets: # remove previous results
                result.destroy()

            i = 0
            for result in self.search_results: # draw new widgets
                song = tk.Label(self.main_canvas, text=self.remove_emoji(result), font=self.controller.song_font, fg="white", bg=self.controller.bg_colour)
                song.bind("<Button-1>", lambda event, song=song: self.test(event, song))

                self.search_result_widgets.append(song)
                self.main_canvas.create_window(self.calc_centre(self.main_canvas), 200+(i*50), window=song)
                i+=1

            self.resize_main_canvas_scroll(None) # pass event as None, update the canvas scrolling function

    def test(self, event, button):
        button.update()
        print(button.winfo_width())

    def remove_emoji(self, string):
        return string.encode('ascii', 'ignore').decode('ascii')

    def calc_centre(self, canvas): # for positioning widgets in the middle of canvas
        canvas.update()
        offset = canvas.winfo_width()//2 # get the midpoint of the screen

        return offset

    # resize functions
    def resize_playlist_canvas_scroll(self, event):
        self.playlist_canvas.configure(scrollregion=self.playlist_canvas.bbox('all'))

    def resize_main_canvas_scroll(self, event):
        self.main_canvas.configure(scrollregion=self.main_canvas.bbox('all'))

    # functions for highlighting animation
    def highlight(self, event, widget):
        self.current_hovered_widget = widget["fg"]
        widget.configure(fg="white")

    def unhighlight(self, event, widget):
        widget.configure(fg=self.current_hovered_widget)

    def main_canvas_yview(self, *args):
        if self.main_canvas.yview() != (0.0, 1.0):
            self.main_canvas.yview(*args)

    def playlist_canvas_yview(self, *args):
        if self.playlist_canvas.yview() != (0.0, 1.0):
            self.playlist_canvas.yview(*args)

    
        

